distributed/spill.py

```diff
@@ -49,28 +82,67 @@ def disk(self) -> Mapping[Hashable, Any]:
         """
         return self.slow
 
-    @staticmethod
-    def _weight(key: Hashable, value: Any) -> int:
-        return safe_sizeof(value)
-
-    def _on_evict(self, key: Hashable, value: Any) -> None:
-        b = safe_sizeof(value)
-        self.spilled_by_key[key] = b
-        self.spilled_total += b
-
-    def _on_retrieve(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key)
-
-    def __setitem__(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
-        super().__setitem__(key, value)
-        if key in self.slow:
-            # value is individually larger than target so it went directly to slow.
-            # _on_evict was not called.
-            b = safe_sizeof(value)
-            self.spilled_by_key[key] = b
-            self.spilled_total += b
-
-    def __delitem__(self, key: Hashable) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
+    @property
+    def spilled_total(self) -> int:
+        return self.slow.total_weight
+
+
+def _in_memory_weight(key: Hashable, value: Any) -> int:
+    return safe_sizeof(value)
+
+
+class MaxSpillExceeded(Exception):
+    pass
+
+
+class PickleError(Exception):
+    pass
+
+
+class Slow(zict.Func):
+    max_weight: int | Literal[False]
+    weight_by_key: dict[Hashable, int]
+    total_weight: int
+
+    def __init__(self, spill_directory: str, max_weight: int | Literal[False] = False):
+        super().__init__(
+            partial(serialize_bytelist, on_error="raise"),
+            deserialize_bytes,
+            zict.File(spill_directory),
+        )
+        self.max_weight = max_weight
+        self.weight_by_key = {}
+        self.total_weight = 0
+
+    def __setitem__(self, key, value):
+        try:
+            pickled = self.dump(value)
+            pickled_size = sum(len(frame) for frame in pickled)
+        except Exception as e:  # This is the error raised when fail to serialize
+            raise PickleError(
+                key, e
+            )  # It's ok to loose the bad key, cluster remains usable
+
+        if (
+            self.max_weight is not False
+            and self.total_weight + pickled_size > self.max_weight
+        ):
+            # Stop callbacks and ensure that the key ends up in SpillBuffer.fast
+            self.total_weight -= self.weight_by_key.pop(key, 0)
+            self.d.pop(key, None)
+            # To be caught by SpillBuffer.__setitem__
+            raise MaxSpillExceeded()
+
+        assert key not in self.weight_by_key  # Thanks to Buffer.__setitem__
+
+        try:
+            self.d[key] = pickled  # pickle and store to disk through File
+        except OSError:  # need to catch OSErrors when writing directly to disk
+            raise
```

Comment:
```suggestion
        # Store to disk through File.
        # This may raise OSError, which is caught by SpillBuffer above.
        self.d[key] = pickled
```

Reply:
@crusaderky  Without the change I did, the OSError doesn't get caught for the case where the key itself is bigger than the target spill. I arrived to this conclusion, after I saw that the change back to what you propose, gives the error reported here https://github.com/dask/distributed/pull/5543#discussion_r783385156

Without that change, when this `self.d[key] = pickled` fails, the exception is not getting caught unless it happens on a callback. You can check this behavior by reversing the change and running the `test_spillbuffer_oserror`

Reply:
Ok, I figured the problem was having this two lines before adding to the dictionary, 
```python
        self.weight_by_key[key] = pickled_size
        self.total_weight += pickled_size
```
fixed it.